docs/streaming-programming-guide.md

```diff
@@ -644,17 +644,90 @@ methods for creating DStreams from files as input sources.
+      time *does not change*.
+    * When `OutputStream.close()` is called, all remaining data is written, the file closed and
+      the NameNode updated with the final size of the file. The modification time is set to
+      the time the file was closed.
+    * File opened for appends via an `append()` operation. This does not change the modification
+      time of the file until the `close()` call is made on the output stream.
+    * `FileSystem.setTimes()` can be used to explicitly set the time on a file.  
+    * The rarely used operations:  `FileSystem.concat()`, `createSnapshot()`, `createSymlink()` and
+      `truncate()` all update the modification time.  
+ 
+    Together, this means that when a file is opened, even before data has been completely written,
+    it may be included in the DStream -after which updates to the file within the same window
+    will be ignored. That is: changes may be missed, and data omitted from the stream. 
+    To guarantee that changes are picked up in a window, write the file
+    to an unmonitored directory, then immediately after the output stream is closed,
+    rename it into the destination directory. 
+    Provided the renamed file appears in the scanned destination directory during the window
+    of its creation, the new data will be picked up.
+
+    Object stores have a different set of limitations.
+
+     + Wildcard directory enumeration can very slow, especially if there are many directories
+       or files to scan.
+     * The file only becomes visible at the end of the write operation; this also defines.
+       the creation time of the file.
+     + A file's  modification time is always the same as its creation time.
+     + The `FileSystem.setTimes()` command to set file timestamps will be ignored.
+     + `FileSystem.rename(file)` of a single file will update the modification time.
+       The time to rename a file is generally `O(length(file))`: the bigger the file, the longer
+       it takes. 
+     + `FileSystem.rename(directory)` is not atomic, and slower the more data there is to rename. 
+       It may update the modification times of renamed files.
+     + Object creation directly though a PUT operation is atomic, irrespective of
+       the programming language or library used to upload the data.
+     + Writing a file to an object store using Hadoop's APIs is an atomic operation;
+       the object is only created via a PUT operation in the final `OutputStream.close()` call.
+
+    Applications using an object store as the direct destination of data
+    should use PUT operations to directly publish data for a DStream to pick up —even though this
+    is not the mechanism to use when writing to HDFS or other filesystem. When using the Hadoop
+    FileSystem API in Spark that means: write the data directory directly to the target directory,
+    knowing that it is the final `close()` call will make the file visible and set its creation &
```

Comment:
I've pulled the details on timestamps into the hadoop docs, and cut back to something that says "filesystems: write then rename, object stores: write directly". That is all people really need to know